c7n/resources/s3.py

```diff
@@ -3707,3 +3708,118 @@ def process_bucket(self, b):
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: bucket-replication
+                    state: True
+                    destination: same-region
+              - name: s3-bucket-has-replication
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: bucket-replication
+                    state: True
+    """
+    schema = type_schema('bucket-replication',
+                         state={'type': 'boolean'},
+                         destination={'type': 'string', 'enum': ['cross-region', 'same-region']})
+
+    permissions = ("s3:GetReplicationConfiguration",)
+    annotation_key = 'Replication'
+
+    def process(self, buckets, event=None):
+        results = []
+        with self.executor_factory(max_workers=2) as w:
+            futures = {w.submit(self.process_bucket, b): b for b in buckets}
+            for future in as_completed(futures):
+                b = futures[future]
+                if future.exception():
+                    self.log.error("Message: %s Bucket: %s", future.exception(),
+                                   b['Name'])
+                    continue
+                if future.result():
+                    results.append(b)
+        return results
+
+    def process_bucket(self, b):
+        client = bucket_client(local_session(self.manager.session_factory), b)
+        rules = []
+        if self.annotation_key not in b:
+            try:
+                bucket_replication = client.get_bucket_replication(Bucket=b['Name'])
+            except ClientError as e:
+                if e.response['Error']['Code'] != 'ReplicationConfigurationNotFoundError':
+                    raise
+                bucket_replication = {}
+            b[self.annotation_key] = bucket_replication
+        else:
+            bucket_replication = b[self.annotation_key]
+
+        rules = bucket_replication.get('ReplicationConfiguration', {}).get('Rules', [])
+        destination = self.data.get('destination')
+        # default `state` to True as previous impl assumed state == True
+        # to preserve backwards compatibility
+        if self.data.get('state', True):
+            if not destination:
+                if rules:
+                    return True
+                else:
+                    return False
+            else:
+                for replication in rules:
+                    if self.filter_bucket(b, replication, destination, client):
+                        return True
+                return False
+        else:
+            if not destination:
+                if rules:
+                    return False
+                else:
+                    return True
+            else:
+                for replication in rules:
+                    if not self.filter_bucket(b, replication, destination, client):
+                        return True
+                return False
```

Comment:
The amount of code paths and nesting here was a bit tricky to follow at first, but it all looks sound. I think it might be a bit clearer if we restructure the conditions, something like:

```suggestion
        state = self.data.get('state', True)

        # If we don't care about cross-region vs. same-region,
        # match based on the presence of rules
        if not destination:
            return bool(rules) == state

        # If any replication rule matches our state/destination pair,
        # we can match the resource and stop looking
        for replication in rules:
            if self.filter_bucket(b, replication, destination, client) == state:
                return True
        return False
```

Reply:
Another option might be to use `any()` here, like:

```python
        # If any replication rule matches our state/destination pair,
        # we can match the resource and stop looking
        return any(
            self.filter_bucket(b, replication, destination, client) == state
            for replication in rules
        )
```

But that might be getting too fancy at the _expense_ of clarity? :shrug:

Reply:
@ajkerrigan , Thank your very much for the suggestions! I optimized the code.